src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigIT.java

```diff
@@ -42,89 +42,133 @@
+    final Template inputTemplate = remoteConfig.getTemplate();
+    final Map<String, Parameter> expectedParameters = getParameters();
+    final Map<String, ParameterGroup> expectedParameterGroups = getParameterGroups();
+    final Version expectedVersion = getVersion();
+    inputTemplate.setParameters(expectedParameters)
+            .setParameterGroups(expectedParameterGroups)
+            .setConditions(CONDITIONS)
+            .setVersion(expectedVersion);
+
+    Template validatedTemplate = remoteConfig.validateTemplate(inputTemplate);
+
+    assertEquals(inputTemplate, validatedTemplate);
+  }
+
+  @Test
+  public void testPublishTemplate() throws FirebaseRemoteConfigException {
     // get template to fetch the active template with correct etag
     final Template oldTemplate = remoteConfig.getTemplate();
     final Template inputTemplate = Template.fromJSON(oldTemplate.toJSON());
-    final String versionNumber = oldTemplate.getVersion().getVersionNumber();
-
+    final Map<String, Parameter> parameters = getParameters();
+    final Map<String, ParameterGroup> parameterGroups = getParameterGroups();
+    final Version version = getVersion();
     // modify template
-    inputTemplate.setParameters(PARAMETERS)
-            .setParameterGroups(PARAMETER_GROUPS)
+    inputTemplate.setParameters(parameters)
+            .setParameterGroups(parameterGroups)
             .setConditions(CONDITIONS)
-            .setVersion(VERSION);
-
-    // validate template
-    Template validatedTemplate = remoteConfig.validateTemplate(inputTemplate);
-    assertEquals(inputTemplate.getETag(), validatedTemplate.getETag());
-    assertEquals(PARAMETERS, validatedTemplate.getParameters());
-    assertEquals(PARAMETER_GROUPS, validatedTemplate.getParameterGroups());
-    assertEquals(CONDITIONS, validatedTemplate.getConditions());
-    assertEquals(VERSION, validatedTemplate.getVersion());
+            .setVersion(version);
 
     // publish template
     Template publishedTemplate = remoteConfig.publishTemplate(inputTemplate);
+
     assertNotEquals(inputTemplate.getETag(), publishedTemplate.getETag());
-    assertEquals(PARAMETERS, publishedTemplate.getParameters());
-    assertEquals(PARAMETER_GROUPS, publishedTemplate.getParameterGroups());
+    assertEquals(parameters, publishedTemplate.getParameters());
+    assertEquals(parameterGroups, publishedTemplate.getParameterGroups());
     assertEquals(CONDITIONS, publishedTemplate.getConditions());
-    assertNotEquals(VERSION, publishedTemplate.getVersion());
+    assertNotEquals(version, publishedTemplate.getVersion());
+  }
+
+  @Test
+  public void testGetTemplate() throws FirebaseRemoteConfigException {
+    // get template to fetch the active template with correct etag
+    // modify and publish a known template to test get template operation.
+    final Template oldTemplate = remoteConfig.getTemplate();
+    final Template inputTemplate = Template.fromJSON(oldTemplate.toJSON());
```

Comment:
Are you using this as a way to "clone" the original template? If so might be better to add a `cloneTemplate()` helper method with this logic.

Reply:
Also we probably don't need to clone it for this test case?

Reply:
Good catch. Thanks! This was only needed in the previous design. Removed the cloning from all the tests.
We only need to copy the etag for `testGetTemplateAtVersion()`. I changed the code there to use the constructor, instead.